### PR TITLE
Add support for match statements to parser

### PR DESCRIPTION
**Pull request by charliermarsh**
# Summary

This is a work-in-progress, but it does work thus far for the implemented cases. I'm putting it up now to (1) signal that I'm working on it, and (2) get feedback from anyone else that has explored this and can see problems ahead.

closes #4323
closes #3427

### CODE REVIEW

**Comment by andersk** [comment]
The main challenge here is that `match` and `case` are [soft keywords](https://docs.python.org/3/reference/lexical_analysis.html#soft-keywords). Naïvely adding them as keyword tokens will break parsing of code like

```python
match = re.match(pattern, str)
if match is not None:
    print(match[1])
```

that uses `match` as an identifier.

We typically don’t know whether `match` should be treated as a keyword or an identifier until many tokens later. Some stupid examples I’ve collected where this results in interesting parse tree differences:

```python
match * a + b, c   # ((match * a) + b), c
match * a + b, c:  # match ((*(a + b)), c):

match -a * b + c   # (match - (a * b)) + c
match -a * b + c:  # match ((-a) * b) + c):

match (a).b   # (match(a)).b
match (a).b:  # match ((a).b):

match [a].b   # (match[a]).b
match [a].b:  # match ([a].b):

match [a,].b   # (match[(a,)]).b   (not (match[a]).b)
match [a,].b:  # match ([a].b):    (not match ([(a,)].b):)
```

My rough plan was going to be to start parsing `match * …`, `match(…`, `match[EXPR]…` under the assumption that it’s an expression with a `match` identifier, but start parsing `match + …`, `match - …`, `match[EXPR, …` under the assumption that it’s a `match` statement, so that in either case there should be enough information to convert to the other form later if we find the initial assumption was wrong. (This specific initial classification is important: there’s not enough information to translate in the other direction in either case.)

**Comment by charliermarsh** [comment]
Thanks, this is extremely helpful. (I kind of assumed I was missing something important given how smoothly this went.)

If you’re further along and want to keep pushing on it, I can close and defer to you. Or if you want to collaborate on a branch that works too. Or I can keep trying here based on some of the pointers above. Whatever you prefer.

**Comment by charliermarsh** [comment]
Some discussion from Parso that may or may not end up being relevant here: https://github.com/davidhalter/parso/issues/138#issuecomment-817378192

**Comment by charliermarsh** [comment]
To what degree would getting those LALRPOP changes merge help here? Do we need to consider publishing a fork for now?

**Comment by charliermarsh** [comment]
I also found this interesting on how Python 3.5 and earlier handled async and await as soft keywords via lexer tricks, but I doubt they’d help us here: https://benjam.info/blog/posts/2019-09-18-python-deep-dive-tokenizer/#async-and-await

**Comment by charliermarsh** [comment]
Although… could we have the lexer look ahead? Is that crazy? Isn’t it a match statement IFF the line starts with match and ends in a colon? (Haven’t thought this through, I assume that won’t work.)

**Comment by andersk** [comment]
No I don’t think a lexer trick like that can work:

```python
match()[a:
    b]
match()[a:
    b]:
```

**Comment by charliermarsh** [comment]
Well you’d need to skip over bracketed content, no? I.e., look ahead until you hit a logical newline (not a physical newline)? At which point you’d either have a semicolon or not.

**Comment by charliermarsh** [comment]
Put differently, what I’m asking is whether we could backtrack or lookahead in the lexer (where we have more control) rather than complicating the parser. E.g. I think LALRPOP does some backtracking in its own lexer (https://github.com/lalrpop/lalrpop/blob/b777a6a884052d6283971f71b3ba453aa6d417ab/lalrpop/src/tok/mod.rs#L138), but I’d have to try this myself to have a real opinion on it.

File: compiler/parser/python.lalrpop
Comment (DimitrisJim) [inline compiler/parser/python.lalrpop:397]: isn't `'as'` allowed here? I.e, probably two alternatives where either a single ident is allowed (as you've currently allowed) and a `<pattern:OrPattern> "as" <alias:Identifier>`.

Comment (DimitrisJim) [reply compiler/parser/python.lalrpop:397]: though I haven't tried to add this yet so lalrpop might not like it (though if it parses the current grammar it should parse that)

Comment (charliermarsh) [reply compiler/parser/python.lalrpop:397]: Yes, oops.

Comment (charliermarsh) [reply compiler/parser/python.lalrpop:397]: (It's just one case here, the no-`as` case is a case on `ClosedPattern`. Have it working locally.)

**Comment by charliermarsh** [comment]
This is able to parse all of [`test_patma.py`](https://github.com/python/cpython/blob/60bbed7f174e481d3fc69984430a4667951678b3/Lib/test/test_patma.py#L133).

**Comment by charliermarsh** [comment]
From what I can tell, performance is the ~same or within 1% (for parsing CPython via Ruff).

File: compiler/parser/python.lalrpop
Comment (charliermarsh) [inline compiler/parser/python.lalrpop:914]: I suspect some of this could be made more concise.

**Comment by youknowone** [comment]
@andersk could you review this PR?

File: compiler/parser/python.lalrpop
Comment (youknowone) [inline compiler/parser/python.lalrpop:590]: I am worrying how many variants can be placed here and the next pattern

**Review by youknowone — APPROVED** [review]
not sure this design fits for full-featured match statement support, but it looks good as starting of it.

**Review by DimitrisJim — APPROVED** [review]

File: compiler/parser/src/soft_keywords.rs
Comment (DimitrisJim) [inline compiler/parser/src/soft_keywords.rs:83]: Under the impression the following should work equally well since we don't really care about what the delimiter token is, just if we are in or not in one:

```suggestion
                    let mut nesting = 0;
                    let mut first = true;
                    let mut seen_colon = false;
                    while let Some(Ok((_, tok, _))) = self.underlying.peek() {
                        match tok {
                            Tok::Newline => break,
                            Tok::Colon if nesting == 0 => {
                                if !first {
                                    seen_colon = true;
                                }
                            }
                            Tok::Lpar | Tok::Lsqb | Tok::Lbrace => nesting += 1,
                            Tok::Rpar | Tok::Rsqb | Tok::Rbrace => nesting -= 1,
                            _ => {}
                        }
                        first = false;
                    }
```

Comment (charliermarsh) [reply compiler/parser/src/soft_keywords.rs:83]: Oh, we may also need to handle unbalanced parens, since we're not guaranteed to get a valid stream here, right? Do unbalanced parens get handled by the lexer, or fail to parse?

Comment (DimitrisJim) [reply compiler/parser/src/soft_keywords.rs:83]: we do handle them being unbalanced from what I remember, they just might not match (and its up to the parser then to reject, i.e `print("hello"}` is lexed just fine).

Comment (DimitrisJim) [reply compiler/parser/src/soft_keywords.rs:83]: (pretty sure CPython should do the same, they use a single `level` counter to track the nesting)

File: compiler/parser/python.lalrpop
Comment (charliermarsh) [reply compiler/parser/python.lalrpop:590]: Do you mean, due to the complexity of reading / understanding it? Or performance? Or something else?

**Comment by youknowone** [comment]
Totally fine if andersk is busy. I forgot the specific branch, but I found @charliermarsh already tested it with ruff somewhere.

**Comment by youknowone** [comment]
unpinning #2759

**Comment by charliermarsh** [comment]
Yeah I've run that branch over the following codebases and, in each case, verified that the lint violation counts are identical before and after that change:

- Hatch (198 violations)
- Pandas (2766 violations)
- Zulip (151 violations)
- scikit-build-core (9 violations)
- Airflow (11 violations)
- CPython (18367 before, 18533 after; however, one fewer parse error in `test_patma.py`, so more violations are _expected_, since we're now actually analyzing that file)

File: compiler/parser/python.lalrpop
Comment (youknowone) [reply compiler/parser/python.lalrpop:590]: I thought this is arithmetic expressions in patterns and manually covering those. Not sure how it is working, but you say it covers everything, right?

Comment (charliermarsh) [reply compiler/parser/python.lalrpop:590]: Yeah, but only a few specific patterns are allowed: a signed number on its own (so "4" or "-4", but not "+4"), or an addition or subtraction where the second value is complex (so "-4 + 4j" or "4 + 4j" or "-4 - 4j" or "-4 - 4j"). Arbitrary arithmetic is not allowed.

Comment (youknowone) [reply compiler/parser/python.lalrpop:590]: Got it. I was misunderstanding the pattern spec.

Comment (charliermarsh) [reply compiler/parser/python.lalrpop:590]: I do think we could DRY up some of the branches here, probably.

**Comment by DimitrisJim** [comment]
Merging so we don't unnecessarily hold up https://github.com/charliermarsh/ruff/issues/282 too, really big thanks for this @charliermarsh!

**Pull request merged**
